File: generate/checks.go
package generate

import (
	"fmt"
	"math"
	"strconv"
	"strings"

	"github.com/Feresey/mtest/schema"
	mapset "github.com/deckarep/golang-set/v2"
)

const pgCatalogPrefix = "pg_catalog."

// var BaseTypes = map[string]struct{}{
// 	"bool": {},

// 	"int2":    {},
// 	"int4":    {},
// 	"int8":    {},
// 	"float4":  {},
// 	"float8":  {},
// 	"numeric": {},

// 	"uuid":    {},
// 	"bytea":   {},
// 	"bit":     {},
// 	"varbit":  {},
// 	"char":    {},
// 	"varchar": {},
// 	"text":    {},

// 	"date":        {},
// 	"time":        {},
// 	"timetz":      {},
// 	"timestamp":   {},
// 	"timestamptz": {},
// }

// TODO это можно вынести как конфиг или типа луа код

var Aliases = map[string]string{
	"int2":   "int",
	"int4":   "int",
	"int8":   "int",
	"float4": "float",
	"float8": "float",

	"uuid":    "text",
	"bytea":   "text",
	"bit":     "text",
	"varbit":  "text",
	"char":    "text",
	"varchar": "text",

	"timetz":      "time",
	"date":        "datetime",
	"timestamp":   "datetime",
	"timestamptz": "datetime",
}

var Checks = map[string][]string{
	"bool": {"True", "False"},

	"int":  {"0", "-1", "1"},
	"int2": {strconv.Itoa(math.MaxInt16), strconv.Itoa(math.MinInt16)},
	"int4": {strconv.Itoa(math.MaxInt32), strconv.Itoa(math.MinInt32)},
	"int8": {strconv.Itoa(math.MaxInt64), strconv.Itoa(math.MinInt64)},
	// numeric типы с явно указанными precision и scale не могут хранить +-Inf
	"numeric": {"'NaN'::NUMERIC"},
	"float":   {"0", "'NaN'::REAL", "'infinity'::REAL", "'-infinity'::REAL"},

	// нет текстовых типов с длиной меньше 1, а нолик для любого текстового типа валидный (вроде)
	"text": {"''", "' '", "'0'"},

	"datetime": {
		"'epoch'::TIMESTAMP",
		"'infinity'::TIMESTAMP",
		"'-infinity'::TIMESTAMP",
	},
	"time": {
		"'allballs'::TIME",
	},
}

type ColumnChecks struct {
	Values []string
}

func (c *ColumnChecks) AddValues(vals ...string) {
	c.Values = append(c.Values, vals...)
}

func (c *ColumnChecks) AddValuesProcess(f func(string) string, vals ...string) {
	for _, v := range vals {
		c.Values = append(c.Values, f(v))
	}
}

// GetDefaultChecks генерирует дефолтные проверки для всех таблиц.
func (g *Generator) GetDefaultChecks(table schema.Table) Records {
	checks := g.getDefaultTableChecks(table)
	// TODO configure mergeChecks
	records := g.transformChecks(checks, true)
	return records
}

// getDefaultTableChecks генерирует дефолтные проверки для указанной таблицы.
func (g *Generator) getDefaultTableChecks(table schema.Table) map[string]ColumnChecks {
	checks := make(map[string]ColumnChecks, len(table.Columns))

	foreignColumns := mapset.NewThreadUnsafeSet[string]()
	for _, fk := range table.ForeignKeys {
		foreignColumns.Append(fk.Constraint.Columns...)
	}

	for _, col := range table.Columns {
		checks[col.Name] = g.makeChecks(col, foreignColumns)
	}

	return checks
}

func (g *Generator) makeChecks(col schema.Column, foreignCols mapset.Set[string]) ColumnChecks {
	var check ColumnChecks
	attr := col.Attributes

	if !attr.NotNullable {
		check.AddValues("NULL")
	}

	// Если тип не является встроенным в postgresql, то я его не обрабатываю.
	if !strings.HasPrefix(col.Type.String(), pgCatalogPrefix) {
		return check
	}

	// Для FK колонок нельзя делать обычные проверки на значения, т.к. они зависят от других таблиц.
	if foreignCols.Contains(col.Name) {
		return check
	}
	g.getTypeChecks(&check, col.Type)

	// TODO numeric min max
	if col.Attributes.IsNumeric && col.Attributes.NumericPrecision == 0 {
		check.AddValues("'infinity'::NUMERIC", "'-infinity'::NUMERIC")
	}

	// Только если это текстовый тип и он имеет аттрибут CharMaxLength, то надо сгенерить строчку максимальной длины.
	// TODO нужно ли проверять на текстовость?
	if attr.HasCharMaxLength {
		check.AddValuesProcess(
			func(s string) string { return fmt.Sprintf("'%s'", s) },
			strings.Repeat(" ", attr.CharMaxLength),
			strings.Repeat("0", attr.CharMaxLength),
		)
	}

	return check
}

func (g *Generator) transformChecks(
	checks map[string]ColumnChecks,
	mergeChecks bool,
) (res Records) {
	// Объединение проверок отдельных колонок в частичные записи.
	for colName, columnChecks := range checks {
		for _, value := range columnChecks.Values {
			pr := Record{
				Columns: []string{colName},
				Values:  []string{value},
			}
			if mergeChecks {
				res.MergeAdd(pr)
			} else {
				res.Records = append(res.Records, pr)
			}
		}
	}

	return res
}

func (g *Generator) getTypeChecks(check *ColumnChecks, typ *schema.DBType) {
	switch typ.TypType() {
	case schema.DataTypeBase:
		g.baseTypesChecks(check, strings.TrimPrefix(typ.String(), pgCatalogPrefix))
	case schema.DataTypeArray:
		// TODO нужно добавить кучу проверок на разные массивы, например для INT[][]:
		// [None], [None, None], [[None]], [], [[1],[2]], [[1],[None]], [[None], [None]]
		// dims := col.Attributes.ArrayDims
	case schema.DataTypeEnum:
		check.AddValuesProcess(func(s string) string {
			return fmt.Sprintf("'%s'::%s", s, typ.String())
		}, typ.EnumValues...)
	case schema.DataTypeDomain:
		// TODO domain attributes checks
		g.getTypeChecks(check, typ.ElemType)
	case schema.DataTypeComposite,
		schema.DataTypeRange,
		schema.DataTypeMultiRange,
		schema.DataTypePseudo:
	default:
	}
}

func (g *Generator) baseTypesChecks(check *ColumnChecks, typeName string) {
	check.AddValues(Checks[Aliases[typeName]]...)
	check.AddValues(Checks[typeName]...)
}

File: generate/checks/lua/checks.lua
local checks = {}
checks.__index = checks

checks.aliases = {
    int2 = "int",
    int4 = "int",
    int8 = "int",
    float4 = "float",
    float8 = "float",

    uuid = "text",
    bytea = "text",
    bit = "text",
    varbit = "text",
    char = "text",
    varchar = "text",

    timetz = "time",
    date = "datetime",
    timestamp = "datetime",
    timestamptz = "datetime",
}

checks.checks = {
    bool = { "True", "False" },

    int = { "0", "-1", "1" },
    int2 = { "-32768", "32767" },
    int4 = { "-2147483648", "2147483647" },
    int8 = { "-9223372036854775808", "9223372036854775807" },
    -- numeric типы с явно указанными precision и scale не могут хранить +-Inf
    numeric = { "'NaN'::NUMERIC" },
    float = { "0", "'NaN'::REAL", "'infinity'::REAL", "'-infinity'::REAL" },

    -- нет текстовых типов с длиной меньше 1, а нолик почти для любого текстового типа валидный
    text = { "''", "' '", "'0'" },

    datetime = {
        "'epoch'::TIMESTAMP",
        "'infinity'::TIMESTAMP",
        "'-infinity'::TIMESTAMP",
    },
    time = {
        "'allballs'::TIME",
    },
}

checks.col_replaces = {
    -- test.users.col1 = {1,2,3},
}
checks.col_appends = {
    -- test.users.col1 = {1,2,3},
}

checks.type_replaces = {
    -- test.typ1 = {1,2,3},
}
checks.type_appends = {
    -- test.typ1 = {1,2,3},
}

checks.skip_tables = {
    -- "test.users"
}

local function mergeLists(dst, src)
    if not src or not dst then return end
    for _, val in ipairs(src) do
        table.insert(dst, val)
    end
end

function checks:get_type_checks(typename)
    local replaces = self.type_replaces[typename]
    if replaces then
        return replaces
    end
    local res = {}
    local aliased = self.checks[self.aliases[typename]]
    mergeLists(res, aliased)
    local base = self.checks[typename]
    mergeLists(res, base)
    local appends = self.type_appends[typename]
    mergeLists(res, appends)
    return res
end

function checks:get_column_checks(t, column)
    local replaces = self.col_replaces[t.name .. "." .. column.name]
    if replaces then
        return replaces
    end
    local col_checks = {}
    if column.attr.notnull then
        table.insert(col_checks, "NULL")
    end

    local max_char = column.attr.char_max_length
    if max_char then
        table.insert(col_checks, "'" .. string.rep(" ", max_char) .. "'")
        table.insert(col_checks, "'" .. string.rep("0", max_char) .. "'")
    end

    if column.attr.is_numeric then
        -- только для типа NUMERIC без явно заданных precision, scale возможны такие значения
        if column.attr.precision == 0 and column.attr.scale == 0 then
            table.insert(col_checks, "'infinity'::NUMERIC")
            table.insert(col_checks, "'-infinity'::NUMERIC")
        end
    end

    local appends = self.col_appends[t.name .. "." .. column.name]

    local res = {}
    mergeLists(res, col_checks)
    mergeLists(res, checks:get_type_checks(column.type.name))
    mergeLists(res, appends)
    return res
end

function checks:get_table_checks(t)
    if self.skip_tables[t.name] then return end

    local foreign_columns = {}
    for _, foreign in pairs(t.fk) do
        for _, colname in ipairs(foreign.columns) do
            foreign_columns[colname] = true
        end
    end

    local res = {}

    for colname, col in pairs(t.columns) do
        if not foreign_columns[col] then
            res[colname] = checks:get_column_checks(t, col)
        end
    end

    return res
end

return checks
